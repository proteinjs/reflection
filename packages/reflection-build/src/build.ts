import * as path from 'path';
const tsconfig = require('./tsconfigTemplate.json');
import { promisifiedFs } from '@proteinjs/util-node';
import { writeGeneratedIndex } from './codegen/writeGeneratedIndex';

export async function build() {
  const targetDir = process.env.INIT_CWD as string;
  const targetDirTsconfig = path.join(targetDir, 'tsconfig.json');
  const targetDirGenerated = path.join(targetDir, 'generated');
  const generatedIndex = path.join(targetDirGenerated, 'index.ts');
  await updatePackageJson();
  await writeTsconfig();
  await writeGeneratedIndex(targetDir, targetDirGenerated, generatedIndex);

  // TODO save their index location in package.json and pass it in to writeGeneratedIndex above
  async function updatePackageJson() {
    const targetDirPackageJson = path.join(targetDir, 'package.json');
    if (!(await promisifiedFs.exists(targetDirPackageJson)))
      throw new Error(`package.json does not exist, run \`npm init -y\` to create one`);

    const targetPackageJson = require(targetDirPackageJson);
    const targetDirDist = path.join(targetDir, 'dist');
    const targetDirDistGenerated = path.join(targetDirDist, 'generated');
    const generatedIndexJs = path.join(targetDirDistGenerated, 'index.js');
    const generatedIndexDts = path.join(targetDirDistGenerated, 'index.d.ts');
    targetPackageJson.main = `./${path.relative(targetDir, generatedIndexJs)}`;
    targetPackageJson.types = `./${path.relative(targetDir, generatedIndexDts)}`;
    await promisifiedFs.writeFile(targetDirPackageJson, JSON.stringify(targetPackageJson, null, 4));
  }

  async function writeTsconfig() {
    if (await promisifiedFs.exists(targetDirTsconfig)) {
      const existingTsconfig = require(targetDirTsconfig);
      if (
        existingTsconfig.include &&
        !existingTsconfig.include.includes(`./${path.relative(targetDir, generatedIndex)}`)
      ) {
        existingTsconfig.include.push(`./${path.relative(targetDir, generatedIndex)}`);
        await promisifiedFs.writeFile(targetDirTsconfig, JSON.stringify(existingTsconfig, null, 4));
      }
    } else {
      let generatedTsconfig = Object.assign({}, tsconfig);
      generatedTsconfig.include = [`./${path.relative(targetDir, generatedIndex)}`];
      await promisifiedFs.writeFile(targetDirTsconfig, JSON.stringify(generatedTsconfig, null, 4));
    }
  }
}
