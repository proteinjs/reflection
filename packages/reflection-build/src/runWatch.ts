#!/usr/bin/env node

import * as path from 'path';
import { cmd } from '@proteinjs/util-node';

// Determine the paths to gulp and the src/watch.ts
const gulpPath = require.resolve('gulp/bin/gulp.js');
const gulpfilePath = path.resolve(__dirname, './watch.js');

// Run the Gulp script
cmd(gulpPath, ['--gulpfile', gulpfilePath, '--cwd', process.cwd()]);
