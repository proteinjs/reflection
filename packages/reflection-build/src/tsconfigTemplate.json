{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "declaration": true,
    "sourceMap": true,
    "outDir": "./dist/",
    "strict": true,
    "noImplicitAny": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true
  },
  "include": ["index.ts"]
}
