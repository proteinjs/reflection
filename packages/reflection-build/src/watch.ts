import * as gulp from 'gulp';
import { exec } from 'child_process';
import * as path from 'path';

// Utility function to execute shell commands
function executeCommand(command: string, cb: (error?: any) => void): void {
  exec(command, (err, stdout, stderr) => {
    if (err) {
      console.error(err.message);
    }
    if (stderr) {
      console.error(stderr);
    }
    if (stdout) {
      console.log(stdout);
    }
    
    cb();  // Always call the callback to signal task completion
  });
}

// Task to run the reflection-build script
gulp.task('reflection-build', (cb) => {
  const command = `node ${path.resolve(__dirname, './runBuild.js')}`;
  executeCommand(command, cb);
});

// Task to run the TypeScript compiler
gulp.task('tsc', (cb) => {
  const command = 'tsc --project ' + path.resolve(process.cwd(), 'tsconfig.json');
  executeCommand(command, cb);
});

gulp.task('watch', () => {
  const watcher = gulp.watch(
    [path.resolve(process.cwd(), 'src/**/*'), path.resolve(process.cwd(), 'test/**/*'), path.resolve(process.cwd(), 'index.ts')],
    gulp.series('reflection-build', 'tsc')
  );

  watcher.on('error', function(this: typeof watcher, err: any) {
    console.error('Error:', err.toString());
    this.emit('end');  // End the task to allow the watch task to continue
  });
});

gulp.task('default', gulp.series('reflection-build', 'tsc', 'watch'));  // run reflection-build and tsc once at the start
