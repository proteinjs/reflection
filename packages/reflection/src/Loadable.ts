export interface Loadable {}
export const LOADABLE_QUALIFIED_NAME = '@proteinjs/reflection/Loadable';
