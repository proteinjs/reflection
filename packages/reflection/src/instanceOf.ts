type ClassConstructor<T> = new (...args: any[]) => T

export function instanceOf<T>(obj: any, cls: ClassConstructor<T>): boolean {
  return obj.constructor.name === cls.name;
}