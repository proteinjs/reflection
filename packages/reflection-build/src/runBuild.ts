#!/usr/bin/env node

import { build } from './build';

(async function() {
	await build();
})();