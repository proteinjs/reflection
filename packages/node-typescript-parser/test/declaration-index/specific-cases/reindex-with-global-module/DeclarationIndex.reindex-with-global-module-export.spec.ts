import { FileChanges } from '../../../../src';
import { join, resolve } from '../../../testUtilities';

import { DeclarationIndex } from '../../../../src/DeclarationIndex';
import { TypescriptParser } from '../../../../src/TypescriptParser';

describe('DeclarationIndex - specific case "reindex-with-global-module"', () => {

    const rootPath = resolve(
        __dirname, '..', '..', '..', '_workspace', 'declaration-index', 'specific-cases',
        'reindex-with-global-module-export',
    );
    const files = [
        join(rootPath, 'node_modules', '@types', 'some-module', 'index.d.ts'),
        join(rootPath, 'classes.ts'),
    ];

    let declarationIndex: DeclarationIndex;

    beforeEach(async () => {
        declarationIndex = new DeclarationIndex(new TypescriptParser(), rootPath);
        await declarationIndex.buildIndex(files);
    });

    it('should parse the declarations', () => {
        expect(declarationIndex.index['some-module']).toMatchSnapshot();
        expect(declarationIndex.index['SomeInterface']).toMatchSnapshot();
        expect(declarationIndex.index['MyClass']).toMatchSnapshot();
    });

    it(`reindex local file doesn't throw error`, async () => {
        const changes: FileChanges = {
            created: [],
            deleted: [],
            updated: [files[1]],
        };
        await declarationIndex.reindexForChanges(changes);
    });

});
