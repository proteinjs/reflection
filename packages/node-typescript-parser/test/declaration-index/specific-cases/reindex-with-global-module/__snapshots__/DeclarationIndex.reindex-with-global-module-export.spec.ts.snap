// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeclarationIndex - specific case "reindex-with-global-module" should parse the declarations 1`] = `
Array [
  DeclarationInfo {
    "declaration": ModuleDeclaration {
      "end": 206,
      "name": "someModule",
      "start": 74,
    },
    "from": "some-module",
  },
]
`;

exports[`DeclarationIndex - specific case "reindex-with-global-module" should parse the declarations 2`] = `
Array [
  DeclarationInfo {
    "declaration": InterfaceDeclaration {
      "accessors": Array [],
      "end": 70,
      "isExported": true,
      "methods": Array [],
      "name": "SomeInterface",
      "properties": Array [],
      "start": 36,
    },
    "from": "some-module",
  },
]
`;

exports[`DeclarationIndex - specific case "reindex-with-global-module" should parse the declarations 3`] = `
Array [
  DeclarationInfo {
    "declaration": ClassDeclaration {
      "accessors": Array [],
      "end": 61,
      "isExported": true,
      "methods": Array [
        MethodDeclaration {
          "end": 59,
          "isAbstract": false,
          "isAsync": false,
          "isOptional": false,
          "isStatic": false,
          "name": "doSomething",
          "parameters": Array [],
          "start": 29,
          "type": "void",
          "variables": Array [],
          "visibility": 2,
        },
      ],
      "name": "MyClass",
      "properties": Array [],
      "start": 1,
    },
    "from": "/classes",
  },
]
`;
